src/snowcli/cli/nativeapp/manager.py

```diff
@@ -97,38 +194,128 @@ def sync_deploy_root_with_stage(self, role: str, app_pkg: str):
+                raise InvalidPackageScriptError(e.name, e)
+
+            except jinja2.UndefinedError as e:
+                raise InvalidPackageScriptError(relpath, e)
+
+        # once we're sure all the templates expanded correctly, execute all of them
+        for queries in queued_queries:
+            self._execute_queries(queries)
 
-        project_name = clean_identifier(self.definition["name"])
-        app_pkg = self.definition.get("package", {}).get(
-            "name", default_app_package(project_name)
+    def _create_dev_app(self, diff: DiffResult) -> None:
+        """
+        (Re-)creates the application with our up-to-date stage.
+        """
+        app_role = (
+            self.definition.get("application", {}).get("role", None) or default_role()
+        )
+
+        with self.use_role(app_role):
+            if self.warehouse:
+                self._execute_query(f"use warehouse {self.warehouse}")
+
+            show_app_cursor = self._execute_query(
+                f"show applications like '{self.app_name}'", cursor_class=DictCursor
+            )
+
+            if show_app_cursor.rowcount != 0:
+                # There can only be one possible pre-existing app pkg with the same name
+                show_app_row: dict = show_app_cursor.fetchone()
+                if (
+                    show_app_row[COMMENT_COL] != SPECIAL_COMMENT
+                    or show_app_row[VERSION_COL] != LOOSE_FILES_MAGIC_VERSION
+                ):
+                    raise ApplicationAlreadyExistsError(self.app_name)
+
+                actual_owner = show_app_row[OWNER_COL]
+                if actual_owner != app_role:
+                    raise UnexpectedOwnerError(self.app_name, app_role, actual_owner)
+
+                if not diff.has_changes():
+                    # the app already exists and is up-to-date
+                    # ensure debug_mode is up-to-date
+                    self._execute_query(
+                        f"alter application {self.app_name} set debug_mode = {self.debug_mode}"
+                    )
+                    return
+                else:
+                    # the app needs to be re-created; drop it
+                    self._execute_query(f"drop application {self.app_name}")
+
+            # Create an app using "loose files" / stage dev mode.
+            log.info(f"Creating new application {self.app_name} in account.")
+            self._execute_query(
+                f"""
+                create application {self.app_name}
+                    from application package {self.package_name}
+                    using @{self.stage_fqn}
+                    debug_mode = {self.debug_mode}
+                    comment = {SPECIAL_COMMENT}
+                """,
+            )
+
+    def app_run(self) -> None:
+        """
+        Implementation of the "snow app run" dev command.
+        """
+        app_pkg_role = (
+            self.definition.get("package", {}).get("role", None) or default_role()
```

Comment:
By moving `default_role()` out of the `.get(...)` call, we short-circuit and can avoid a round-trip to Snowflake when we've set the role in the project definition.